File: yntra-ios/YntraIOS/Views/JobsView.swift
import SwiftUI
import yntra_core

struct JobsView: View {
    @ObservedObject var viewModel: JobsViewModel
    
    @State private var selectedJob: JobTicket? = nil
    
    private let darkBackground = Color(red: 0.04, green: 0.06, blue: 0.1)
    private let cardBackground = Color(red: 0.12, green: 0.16, blue: 0.23)

    var body: some View {
        ZStack {
            darkBackground.ignoresSafeArea()
            
            VStack(alignment: .leading, spacing: 20) {
                // Header
                HStack {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("Work Orders")
                            .font(.system(size: 28, weight: .black, design: .rounded))
                            .foregroundColor(.white)
                        Text("Logistics & Move Management")
                            .font(.system(size: 13, weight: .medium, design: .rounded))
                            .foregroundColor(.white.opacity(0.6))
                    }
                    Spacer()
                }
                .padding(.horizontal)
                .padding(.top, 10)
                
                if let errorMessage = viewModel.errorMessage {
                    Text("Error: \(errorMessage)")
                        .font(.caption)
                        .foregroundColor(.red)
                        .padding(.horizontal)
                }
                
                if viewModel.jobs.isEmpty {
                    VStack(spacing: 12) {
                        Spacer()
                        Image(systemName: "shippingbox.fill")
                            .font(.system(size: 40))
                            .foregroundColor(.white.opacity(0.3))
                        Text("No job tickets assigned")
                            .font(.system(size: 14, weight: .medium, design: .rounded))
                            .foregroundColor(.white.opacity(0.5))
                        Spacer()
                    }
                    .frame(maxWidth: .infinity)
                } else {
                    ScrollView {
                        LazyVStack(spacing: 12) {
                            ForEach(viewModel.jobs, id: \.id) { job in
                                Button(action: {
                                    selectedJob = job
                                }) {
                                    VStack(alignment: .leading, spacing: 10) {
                                        HStack {
                                            Text(job.title)
                                                .font(.system(size: 16, weight: .bold, design: .rounded))
                                                .foregroundColor(.white)
                                                .multilineTextAlignment(.leading)
                                            
                                            Spacer()
                                            
                                            StatusBadgeView(status: job.status)
                                        }
                                        
                                        Text(job.description)
                                            .font(.system(size: 13, weight: .medium, design: .rounded))
                                            .foregroundColor(.white.opacity(0.6))
                                            .lineLimit(2)
                                            .multilineTextAlignment(.leading)
                                        
                                        Divider()
                                            .background(Color.white.opacity(0.1))
                                        
                                        HStack {
                                            HStack(spacing: 6) {
                                                Image(systemName: "mappin.and.ellipse")
                                                    .font(.system(size: 14))
                                                    .foregroundColor(Color(red: 0.55, green: 0.36, blue: 0.96))
                                                Text(job.locationAddress)
                                                    .font(.system(size: 12, weight: .medium, design: .rounded))
                                                    .foregroundColor(.white.opacity(0.8))
                                            }
                                            
                                            Spacer()
                                            
                                            Text(job.scheduledDate)
                                                .font(.system(size: 12, weight: .bold, design: .rounded))
                                                .foregroundColor(Color(red: 0.55, green: 0.36, blue: 0.96))
                                        }
                                    }
                                    .padding(18)
                                    .background(cardBackground)
                                    .cornerRadius(18)
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 18)
                                            .stroke(Color.white.opacity(0.05), lineWidth: 1)
                                    )
                                }
                            }
                        }
                        .padding(.horizontal)
                    }
                }
            }
        }
        .sheet(item: $selectedJob) { job in
            JobDetailView(job: job, viewModel: viewModel)
        }
    }
}

struct StatusBadgeView: View {
    let status: String
    
    private var badgeColors: (bg: Color, text: Color) {
        switch status {
        case "assigned":
            return (Color(red: 0.99, green: 0.95, blue: 0.78), Color(red: 0.85, green: 0.47, blue: 0.02))
        case "in_progress":
            return (Color(red: 0.86, green: 0.92, blue: 1.0), Color(red: 0.15, green: 0.39, blue: 0.92))
        case "completed":
            return (Color(red: 0.82, green: 0.98, blue: 0.9), Color(red: 0.02, green: 0.59, blue: 0.41))
        default:
            return (Color(red: 0.95, green: 0.96, blue: 0.96), Color(red: 0.29, green: 0.33, blue: 0.39))
        }
    }
    
    var body: some View {
        Text(status.replacingOccurrences(of: "_", with: " ").uppercased())
            .font(.system(size: 10, weight: .black, design: .rounded))
            .foregroundColor(badgeColors.text)
            .padding(.horizontal, 8)
            .padding(.vertical, 4)
            .background(badgeColors.bg)
            .cornerRadius(8)
    }
}

struct JobDetailView: View {
    let job: JobTicket
    @ObservedObject var viewModel: JobsViewModel
    @Environment(\.presentationMode) var presentationMode
    
    @State private var checklist: [ChecklistTaskItem] = []
    @State private var reportText: String = ""
    
    private let darkBackground = Color(red: 0.04, green: 0.06, blue: 0.1)
    private let cardBackground = Color(red: 0.12, green: 0.16, blue: 0.23)
    
    var body: some View {
        NavigationView {
            ZStack {
                darkBackground.ignoresSafeArea()
                
                ScrollView {
                    VStack(alignment: .leading, spacing: 20) {
                        VStack(alignment: .leading, spacing: 6) {
                            Text(job.title)
                                .font(.title3)
                                .fontWeight(.bold)
                                .foregroundColor(.white)
                            
                            HStack {
                                StatusBadgeView(status: job.status)
                                Spacer()
                                Text(job.scheduledDate)
                                    .font(.caption)
                                    .foregroundColor(.white.opacity(0.6))
                            }
                        }
                        
                        Text(job.description)
                            .font(.system(size: 14, weight: .medium, design: .rounded))
                            .foregroundColor(.white.opacity(0.8))
                        
                        Divider()
                            .background(Color.white.opacity(0.1))
                        
                        // Route Details
                        if job.originAddress != nil || job.destinationAddress != nil {
                            VStack(alignment: .leading, spacing: 8) {
                                Text("ROUTE DETAILS")
                                    .font(.system(size: 11, weight: .bold))
                                    .foregroundColor(.white.opacity(0.5))
                                
                                if let origin = job.originAddress {
                                    Text("Origin: \(origin) (Floor: \(job.originFloor), Elevator: \(job.originHasElevator ? "Yes" : "No"))")
                                        .font(.system(size: 13, weight: .medium, design: .rounded))
                                        .foregroundColor(.white)
                                }
                                
                                if let dest = job.destinationAddress {
                                    Text("Destination: \(dest) (Floor: \(job.destinationFloor), Elevator: \(job.destinationHasElevator ? "Yes" : "No"))")
                                        .font(.system(size: 13, weight: .medium, design: .rounded))
                                        .foregroundColor(.white)
                                }
                            }
                            
                            Divider()
                                .background(Color.white.opacity(0.1))
                        }
                        
                        // Checklist Section
                        VStack(alignment: .leading, spacing: 10) {
                            Text("CHECKLIST / TASKS")
                                .font(.system(size: 11, weight: .bold))
                                .foregroundColor(.white.opacity(0.5))
                            
                            if checklist.isEmpty {
                                Text("No checklist items.")
                                    .font(.system(size: 13, weight: .medium, design: .rounded))
                                    .foregroundColor(.white.opacity(0.5))
                            } else {
                                ForEach(0..<checklist.count, id: \.self) { index in
                                    HStack(spacing: 12) {
                                        Button(action: {
                                            if job.status == "in_progress" {
                                                checklist[index].done.toggle()
                                            }
                                        }) {
                                            Image(systemName: checklist[index].done ? "checkmark.square.fill" : "square")
                                                .font(.system(size: 20))
                                                .foregroundColor(checklist[index].done ? Color(red: 0.55, green: 0.36, blue: 0.96) : .white.opacity(0.5))
                                        }
                                        .disabled(job.status != "in_progress")
                                        
                                        Text(checklist[index].text)
                                            .font(.system(size: 13, weight: .medium, design: .rounded))
                                            .foregroundColor(checklist[index].done ? .white.opacity(0.5) : .white)
                                        
                                        Spacer()
                                    }
                                }
                            }
                        }
                        
                        // Completion Report
                        if job.status == "in_progress" {
                            Divider()
                                .background(Color.white.opacity(0.1))
                            
                            VStack(alignment: .leading, spacing: 8) {
                                Text("SUBMIT COMPLETION REPORT")
                                    .font(.system(size: 11, weight: .bold))
                                    .foregroundColor(.white.opacity(0.5))
                                
                                TextField("Describe the completed work...", text: $reportText)
                                    .textFieldStyle(PlainTextFieldStyle())
                                    .padding(12)
                                    .background(cardBackground)
                                    .cornerRadius(10)
                                    .foregroundColor(.white)
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 10)
                                            .stroke(Color.white.opacity(0.1), lineWidth: 1)
                                    )
                            }
                        } else if job.status == "completed" {
                            Divider()
                                .background(Color.white.opacity(0.1))
                            
                            VStack(alignment: .leading, spacing: 8) {
                                Text("COMPLETION REPORT")
                                    .font(.system(size: 11, weight: .bold))
                                    .foregroundColor(.white.opacity(0.5))
                                
                                Text(job.completionReport ?? "No report text submitted.")
                                    .font(.system(size: 13, weight: .medium, design: .rounded))
                                    .foregroundColor(.white.opacity(0.8))
                            }
                        }
                        
                        Spacer(minLength: 40)
                        
                        // Action Buttons
                        if job.status == "assigned" {
                            Button(action: {
                                viewModel.updateStatus(jobId: job.id, status: "in_progress")
                                presentationMode.wrappedValue.dismiss()
                            }) {
                                Text("Start Job")
                                    .foregroundColor(.white)
                                    .frame(maxWidth: .infinity)
                                    .padding()
                                    .background(Color(red: 0.55, green: 0.36, blue: 0.96))
                                    .cornerRadius(12)
                            }
                        } else if job.status == "in_progress" {
                            Button(action: {
                                let encoder = JSONEncoder()
                                if let data = try? encoder.encode(checklist),
                                   let jsonString = String(data: data, encoding: .utf8) {
                                    viewModel.completeJob(jobId: job.id, checklistJson: jsonString, completionReport: reportText)
                                }
                                presentationMode.wrappedValue.dismiss()
                            }) {
                                Text("Complete & Submit")
                                    .foregroundColor(.white)
                                    .frame(maxWidth: .infinity)
                                    .padding()
                                    .background(Color(red: 0.1, green: 0.73, blue: 0.51))
                                    .cornerRadius(12)
                            }
                        }
                    }
                    .padding()
                }
            }
            .navigationBarTitleDisplayMode(.inline)
            .navigationBarItems(trailing: Button("Close") {
                presentationMode.wrappedValue.dismiss()
            }.foregroundColor(.white))
            .onAppear {
                parseChecklist()
            }
        }
    }
    
    private func parseChecklist() {
        guard let data = job.checklistJson.data(using: .utf8) else { return }
        let decoder = JSONDecoder()
        if let decoded = try? decoder.decode([ChecklistTaskItem].self, from: data) {
            self.checklist = decoded
        }
    }
}

struct ChecklistTaskItem: Codable, Identifiable {
    var id = UUID()
    let text: String
    var done: Bool
    
    enum CodingKeys: String, CodingKey {
        case text, done
    }
}

extension JobTicket: Identifiable {}

File: yntra-android/app/src/main/java/com/yntra/app/views/JobsView.kt
package com.yntra.app.views

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.yntra.app.viewmodels.JobsViewModel
import uniffi.yntra_core.JobTicket
import org.json.JSONArray
import org.json.JSONObject

data class ChecklistTask(val text: String, val done: Boolean)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun JobsView(viewModel: JobsViewModel) {
    val jobs by viewModel.jobs.collectAsState()
    val errorMessage by viewModel.errorMessage.collectAsState()

    var selectedJob by remember { mutableStateOf<JobTicket?>(null) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFF0B0F19))
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Text(
            text = "Logistics Work Orders",
            color = Color.White,
            fontSize = 26.sp,
            fontWeight = FontWeight.Black
        )

        errorMessage?.let {
            Card(
                colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.errorContainer),
                shape = RoundedCornerShape(12.dp)
            ) {
                Text(
                    text = it,
                    color = MaterialTheme.colorScheme.onErrorContainer,
                    modifier = Modifier.padding(12.dp),
                    fontSize = 14.sp
                )
            }
        }

        if (jobs.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .background(Color(0xFF1E293B), RoundedCornerShape(16.dp)),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.Build,
                        contentDescription = "Empty",
                        tint = Color(0xFF94A3B8),
                        modifier = Modifier.size(36.dp)
                    )
                    Text(text = "No job tickets assigned", color = Color(0xFF94A3B8), fontSize = 14.sp)
                }
            }
        } else {
            LazyColumn(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(jobs) { job ->
                    JobTicketCard(job = job, onClick = { selectedJob = job })
                }
            }
        }
    }

    selectedJob?.let { job ->
        JobDetailDialog(
            job = job,
            viewModel = viewModel,
            onDismiss = { selectedJob = null }
        )
    }
}

@Composable
fun JobTicketCard(job: JobTicket, onClick: () -> Unit) {
    Card(
        colors = CardDefaults.cardColors(containerColor = Color(0xFF1E293B)),
        shape = RoundedCornerShape(16.dp),
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = job.title,
                    color = Color.White,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.weight(1f)
                )
                
                Spacer(modifier = Modifier.width(8.dp))
                StatusBadge(status = job.status)
            }

            Text(
                text = job.description,
                color = Color(0xFF94A3B8),
                fontSize = 13.sp,
                maxLines = 2
            )

            Divider(color = Color(0xFF334155), thickness = 1.dp)

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Icon(
                        imageVector = Icons.Default.LocationOn,
                        contentDescription = "Address",
                        tint = Color(0xFF8B5CF6),
                        modifier = Modifier.size(16.dp)
                    )
                    Text(
                        text = job.locationAddress,
                        color = Color(0xFFCBD5E1),
                        fontSize = 12.sp
                    )
                }

                Text(
                    text = job.scheduledDate,
                    color = Color(0xFF8B5CF6),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

@Composable
fun StatusBadge(status: String) {
    val (bgColor, textColor) = when (status) {
        "assigned" -> Color(0xFFFEF3C7) to Color(0xFFD97706)
        "in_progress" -> Color(0xFFDBEAFE) to Color(0xFF2563EB)
        "completed" -> Color(0xFFD1FAE5) to Color(0xFF059669)
        else -> Color(0xFFF3F4F6) to Color(0xFF4B5563)
    }

    Surface(
        color = bgColor,
        shape = RoundedCornerShape(8.dp)
    ) {
        Text(
            text = status.uppercase().replace("_", " "),
            color = textColor,
            fontSize = 10.sp,
            fontWeight = FontWeight.Black,
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
        )
    }
}

@Composable
fun JobDetailDialog(
    job: JobTicket,
    viewModel: JobsViewModel,
    onDismiss: () -> Unit
) {
    var reportText by remember { mutableStateOf("") }
    val tasks = remember(job.checklistJson) {
        val list = mutableStateListOf<ChecklistTask>()
        try {
            val jsonArr = JSONArray(job.checklistJson)
            for (i in 0 until jsonArr.length()) {
                val obj = jsonArr.getJSONObject(i)
                list.add(ChecklistTask(obj.getString("text"), obj.optBoolean("done", false)))
            }
        } catch (e: Exception) {
            // Fallback
        }
        list
    }

    AlertDialog(
        onDismissRequest = onDismiss,
        containerColor = Color(0xFF1E293B),
        title = {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = job.title,
                    color = Color.White,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.weight(1f)
                )
                IconButton(onClick = onDismiss) {
                    Icon(imageVector = Icons.Default.Close, contentDescription = "Close", tint = Color.White)
                }
            }
        },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 8.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Description
                Text(
                    text = job.description,
                    color = Color(0xFFCBD5E1),
                    fontSize = 14.sp
                )

                Divider(color = Color(0xFF334155))

                // Addresses
                if (job.originAddress != null || job.destinationAddress != null) {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        Text(text = "ROUTE DETAILS", color = Color(0xFF94A3B8), fontSize = 11.sp, fontWeight = FontWeight.Bold)
                        Text(
                            text = "Origin: ${job.originAddress ?: "N/A"} (Floor: ${job.originFloor}, Elevator: ${if (job.originHasElevator) "Yes" else "No"})",
                            color = Color(0xFFCBD5E1),
                            fontSize = 12.sp
                        )
                        Text(
                            text = "Destination: ${job.destinationAddress ?: "N/A"} (Floor: ${job.destinationFloor}, Elevator: ${if (job.destinationHasElevator) "Yes" else "No"})",
                            color = Color(0xFFCBD5E1),
                            fontSize = 12.sp
                        )
                    }
                    Divider(color = Color(0xFF334155))
                }

                // Checklist
                Text(text = "CHECKLIST / TASKS", color = Color(0xFF94A3B8), fontSize = 11.sp, fontWeight = FontWeight.Bold)
                if (tasks.isEmpty()) {
                    Text(text = "No checklist items.", color = Color(0xFF94A3B8), fontSize = 13.sp)
                } else {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        tasks.forEachIndexed { index, task ->
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .clickable(enabled = job.status == "in_progress") {
                                        tasks[index] = task.copy(done = !task.done)
                                    },
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Checkbox(
                                    checked = task.done,
                                    enabled = job.status == "in_progress",
                                    onCheckedChange = { isChecked ->
                                        tasks[index] = task.copy(done = isChecked)
                                    },
                                    colors = CheckboxDefaults.colors(
                                        checkedColor = Color(0xFF8B5CF6),
                                        checkmarkColor = Color.White
                                    )
                                )
                                Text(
                                    text = task.text,
                                    color = if (task.done) Color(0xFF94A3B8) else Color.White,
                                    fontSize = 13.sp
                                )
                            }
                        }
                    }
                }

                // Completion Report Section
                if (job.status == "in_progress") {
                    Divider(color = Color(0xFF334155))
                    Text(text = "SUBMIT COMPLETION REPORT", color = Color(0xFF94A3B8), fontSize = 11.sp, fontWeight = FontWeight.Bold)
                    OutlinedTextField(
                        value = reportText,
                        onValueChange = { reportText = it },
                        placeholder = { Text("Describe the completed work...") },
                        modifier = Modifier.fillMaxWidth(),
                        colors = OutlinedTextFieldDefaults.colors(
                            focusedBorderColor = Color(0xFF8B5CF6),
                            unfocusedBorderColor = Color(0xFF334155),
                            focusedTextColor = Color.White,
                            unfocusedTextColor = Color.White,
                            placeholderColor = Color(0xFF94A3B8)
                        )
                    )
                } else if (job.status == "completed") {
                    Divider(color = Color(0xFF334155))
                    Text(text = "COMPLETION REPORT", color = Color(0xFF94A3B8), fontSize = 11.sp, fontWeight = FontWeight.Bold)
                    Text(
                        text = job.completionReport ?: "No report text submitted.",
                        color = Color(0xFFCBD5E1),
                        fontSize = 13.sp
                    )
                }
            }
        },
        confirmButton = {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                if (job.status == "assigned") {
                    Button(
                        onClick = {
                            viewModel.updateStatus(job.id, "in_progress")
                            onDismiss()
                        },
                        colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF8B5CF6))
                    ) {
                        Text("Start Job", color = Color.White)
                    }
                } else if (job.status == "in_progress") {
                    Button(
                        onClick = {
                            // Serialize checklist back to JSON
                            val jsonArr = JSONArray()
                            tasks.forEach { t ->
                                val obj = JSONObject()
                                obj.put("text", t.text)
                                obj.put("done", t.done)
                                jsonArr.put(obj)
                            }
                            viewModel.completeJob(job.id, jsonArr.toString(), reportText)
                            onDismiss()
                        },
                        colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF10B981))
                    ) {
                        Text("Complete & Submit", color = Color.White)
                    }
                }
            }
        }
    )
}
